lib/rspec/sql/query_matcher.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+require_relative "query_summary"
+
+module RSpec
+  module Sql
+    # Compares expectations with actual queries.
+    class QueryMatcher
+      attr_reader :expected
+
+      def initialize(queries, expected)
+        @queries = queries
+        @expected = sanitize_number(expected)
+      end
+
+      def matches?
+        # Simple presence validation.
+        return !@queries.empty? if expected.nil?
+
+        actual == expected
+      end
+
+      def actual
+        @actual ||= actual_compared_to_expected
+      end
+
+      private
+
+      # Support writing: `is_expected.to query_database 5.times`
+      def sanitize_number(expected)
+        if expected.is_a?(Enumerator) && expected.inspect.match?(/:times>$/)
+          expected.size
+        else
+          expected
+        end
+      end
+
+      def actual_compared_to_expected
+        case expected
+        when Integer
+          @queries.size
+        when Array
+          query_names
+        when Hash
+          query_summary
+        else
+          raise "What are you expecting?"
+        end
+      end
+
+      def query_names
+        @queries.map { |q| q[:name] || q[:sql].split.take(2).join(" ") }
+      end
+
+      def query_summary
+        QuerySummary.new(@queries).summary
```

Comment:
To minimise the dependency on `QuerySummary`, I would inject it in the constructor like this : 
```ruby
def initialize(queries, expected, query_summary=QuerySummary)
  ...
end
```
Now you are depending on an object that implement `summary` and not `QuerySummary` directly.

Reply:
I think it's okay to depend on QuerySummary here. It means that the calling module doesn't need to know the QuerySummary. It reduces the API and the matcher definition doesn't need to know which class is responsible for which format.